velox/common/caching/AsyncDataCache.cpp

```diff
@@ -504,4 +505,48 @@ std::string AsyncDataCache::toString() const {
   return out.str();
 }
 
+CoalescedIoStats readPins(
+    const std::vector<CachePin>& pins,
+    int32_t maxGap,
+    int32_t rangesPerIo,
+    std::function<uint64_t(int32_t index)> offsetFunc,
+    std::function<void(
+        const std::vector<CachePin>& pins,
+        int32_t begin,
+        int32_t end,
+        uint64_t offset,
+        const std::vector<folly::Range<char*>>& buffers)> readFunc) {
+  return coalescedIo<CachePin, folly::Range<char*>>(
+      pins,
+      maxGap,
+      rangesPerIo,
+      offsetFunc,
+      [&](int32_t index) { return pins[index].entry()->size(); },
+      [&](const CachePin& pin) {
+        return std::max<int32_t>(1, pin.entry()->data().numRuns());
```

Comment:
Copying comments from the other PR
@oerling this is flaky since the number of runs can vary based on how the malloc allocates the ranges. If the ranges are consecutive, we end up coalescing the runs.
```
void MappedMemory::Allocation::append(uint8_t* address, int32_t numPages)
// Increment page count if new data starts at end of the last run
// and the combined page count is within limits.
  if (address == last.data() + last.numPages() * kPageSize &&
      last.numPages() + numPages <= PageRun::kMaxPagesInRun) {
    runs_.back() = PageRun(last.data(), last.numPages() + numPages);
  } else {
    runs_.emplace_back(address, numPages);
  }
```

Reply:
to clarify, the AsyncDataCacheTest.cpp that depends on this is flaky as the number of runs can vary between 1 to 3.